csrc/ops/arith.cpp

```diff
@@ -278,27 +278,47 @@ TensorView* iota(Val* length, Val* start, Val* step, DataType dtype) {
       "length must be integer, but get dtype ",
       *length->getDataType());
   TORCH_CHECK(
-      isIntegralType(*start->getDataType()) == isIntegralType(dtype) &&
+      !isComplexType(*start->getDataType()) &&
+          isIntegralType(*start->getDataType()) == isIntegralType(dtype) &&
           isFloatingPointType(*start->getDataType()) ==
               isFloatingPointType(dtype),
-      "dtype does not match for iota, should be ",
+      "iota: start dtype does not match specified dtype argument, should be ",
       dtype,
       " but get ",
       *start->getDataType());
   TORCH_CHECK(
-      isIntegralType(*step->getDataType()) == isIntegralType(dtype) &&
+      !isComplexType(*step->getDataType()) &&
+          isIntegralType(*step->getDataType()) == isIntegralType(dtype) &&
           isFloatingPointType(*step->getDataType()) ==
               isFloatingPointType(dtype),
-      "dtype does not match for iota, should be ",
+      "iota: step dtype does not match specified dtype argument, should be ",
       dtype,
       " but get ",
       *step->getDataType());
+
   if (start->getDataType() != dtype) {
     start = castOp(dtype, start);
   }
   if (step->getDataType() != dtype) {
     step = castOp(dtype, step);
   }
+
+  if (start->isConst() && start->isFloatingPointScalar()) {
```

Comment:
You may want to reverse these to avoid calling `isConst` in cases where `start` is not fp.

Reply:
I only wanted the check for floating point numbers, since integers cannot be infinite.